utilities/object_registry.cc

```diff
@@ -26,61 +55,412 @@ const ObjectLibrary::Entry *ObjectLibrary::FindEntry(
   for (const auto &iter : entries_) {
-    ROCKS_LOG_HEADER(logger, "    Registered factories for type[%s] ",
-                     iter.first.c_str());
-    bool printed_one = false;
-    for (const auto &e : iter.second) {
-      ROCKS_LOG_HEADER(logger, "%c %s", (printed_one) ? ',' : ':',
-                       e->Name().c_str());
-      printed_one = true;
+    types->insert(iter.first);
+    count++;
+  }
+  return count;
+}
+
+size_t ObjectLibrary::GetRegisteredNames(
+    const std::string &type, std::vector<std::string> *names) const {
+  size_t count = 0;
+  auto iter = entries_.find(type);
+  if (iter == entries_.end()) {
+    count = 0;  // No entry, no count
+  } else {
+    count = iter->second.size();
+    if (names != nullptr) {
+      for (const auto &f : iter->second) {
+        names->push_back(f->Name());
+      }
     }
   }
-  ROCKS_LOG_HEADER(logger, "\n");
+  return count;
 }
+class DefaultObjectLibrary : public ObjectLibrary {
+ public:
+  const char *Name() const override { return kDefaultLibraryName.c_str(); }
+
+  virtual std::string GetPrintableOptions() const override {
+    return AsPrintableOptions("default_library");
+  }
+};
+
+struct ObjectRegistryOptions {
+  // The set of libraries to search for factories for this registry.
+  // The libraries are searched in reverse order (back to front) when
+  // searching for entries.
+  std::vector<std::shared_ptr<ObjectLibrary>> libraries;
+};
+
+static std::unordered_map<std::string, OptionTypeInfo>
+    object_registry_library_type_info = {
+        {"libraries",
+         OptionTypeInfo::Vector<std::shared_ptr<ObjectLibrary>>(
+             offsetof(struct ObjectRegistryOptions, libraries),
+             OptionVerificationType::kNormal, OptionTypeFlags::kCompareNever,
+             OptionTypeInfo::AsCustomSharedPtr<ObjectLibrary>(
+                 0, OptionVerificationType::kNormal,
+                 OptionTypeFlags::kShared | OptionTypeFlags::kCompareNever))},
+};
 
 // Returns the Default singleton instance of the ObjectLibrary
 // This instance will contain most of the "standard" registered objects
 std::shared_ptr<ObjectLibrary> &ObjectLibrary::Default() {
   static std::shared_ptr<ObjectLibrary> instance =
-      std::make_shared<ObjectLibrary>();
+      std::make_shared<DefaultObjectLibrary>();
   return instance;
 }
 
+class ObjectRegistryImpl : public ObjectRegistry {
+ public:
+  ObjectRegistryImpl() {
+    options_.libraries.push_back(ObjectLibrary::Default());
+    ConfigurableHelper::RegisterOptions(*this, "Libraries", &options_,
+                                        &object_registry_library_type_info);
+  }
+
+  ObjectRegistryImpl(const ObjectRegistryImpl &that) : options_(that.options_) {
+    ConfigurableHelper::RegisterOptions(*this, "Libraries", &options_.libraries,
+                                        &object_registry_library_type_info);
+  }
+
+  // Makes a copy of this registry.
+  std::shared_ptr<ObjectRegistry> Clone() const override {
```

Comment:
I'm not sure `Clone` is a good function name here. We just share the same OBjectRegistry instance pointer right? Not clone a new instance.